Adds support for site app catalog for app get, list, add, remove and deploy

VelinGeorgiev opened the pull request:
Adds support for site app catalog for app get, list, add, remove and deploy solving part 1 from #405 

This will allow users to deploy and remove apps with skip feature deployment and I will complete the rest of the commands (install, retract, uninstall, upgrade) in the following weeks. 

@waldekmastykarz, please do not forget to give the credit to @andrewconnell  :)

waldekmastykarz commented:
Thanks! Will definitely do!

waldekmastykarz commented on docs/manual/docs/cmd/spo/app/app-add.md:
````diff
@@ -24,26 +25,33 @@ Option|Description
+spo app add --filePath sharepoint/solution/spfx.sppkg --overwrite
 ```
 
+Add the _spfx.sppkg_ package to the site collection app catalog of the site you are currencly logged in
````
In the CLI we don't use the concept of current site. When we have commands that operate on a particular site, we ask the user to specify the URL in the command arguments. Because authentication is typically interactive, the current format of the command would be impractical, because it would require you to reconnect in the middle of the script if you wanted to add a solution to site collection app catalogs of two different sites. Could we change this command to allow specifying the URL of the site where the site collection app catalog is located? Additionally, could you double check that all other commands follow this pattern?

VelinGeorgiev replied:
Yes, I will change that shortly. Should I not allow update if site collection app catalog specified, but no siteUrl specified?

waldekmastykarz commented:
Machine! I'll have a look shortly. Thank you! 👏

VelinGeorgiev commented:
Makes sense and it will be more clean.  @waldekmastykarz,  I can do that shortly.

VelinGeorgiev commented:
@waldekmastykarz, siteUrl is replaced now with appCatalogUrl in the last commit https://github.com/pnp/office365-cli/pull/653/commits/6554ca12f4b698499e1b6693fe60c4d1480b1dbf

waldekmastykarz commented:
Going through your latest changes, I realized something pretty confusing as a result of my suggestion to change `siteUrl` to `appCatalogUrl`. In other commands, where we already had `appCatalogUrl` it was meant to point to the tenant app catalog site, not the actual app catalog list. In the context of a site collection app catalog however, this is confusing, because the name `appCatalogUrl` indicates, as you correctly assumed, the URL of the catalog list, not the URL of the site. But, because the app catalog is always called _AppCatalog_, there is no point in including it in the URL and we don't need it for anything (we're actually removing it from the URL). So thinking about it, I'd say, we should go back to your original suggestion and use `siteUrl` instead.
Terribly sorry for the trouble and not thinking it through originally. I'll adjust it myself when merging the PR as I don't want to cause you any more work again, but just wanted to let you know why things will be different in the merged code and avoid any surprises.

waldekmastykarz commented on src/o365/spo/commands/app/app-add.ts:
```diff
@@ -67,13 +64,21 @@ class SpoAppAddCommand extends SpoCommand {
         const fileName: string = path.basename(fullPath);
         const requestOptions: any = {
-          url: `${auth.site.url}/_api/web/tenantappcatalog/Add(overwrite=${(overwrite.toString().toLowerCase())}, url='${fileName}')`,
+          url: `${appCatalogUrl}/_api/web/${scope}appcatalog/Add(overwrite=${(overwrite.toString().toLowerCase())}, url='${fileName}')`,
```
When connected to the tenant admin site, this command fails, because here we're using the URL of the site passed in the command options, while on line 80 we're using the access token from the site used to log in.

waldekmastykarz commented:
Thank you @VelinGeorgiev!

VelinGeorgiev commented:
@waldekmastykarz I had to do small refactoring to make sure the right site has the right accessToken. As a result of my refactoring:
- I've created an app-base.ts class with shared app methods.
- I've moved one method from SpoCommand to the app base class since it was used only by the app commands
- Search query to find app catalog in the getTenantAppCatalogUrl method is replaced by client.svc query .This removes edge cases where the app catalog is not indexed or has set NoCrawl to true. Having that in place removes the need of additional prompt when search does not find the app catalog and I removed it.

waldekmastykarz commented:
Thanks! Have you considered using the REST API to get the tenant app catalog URL as I described here: https://blog.mastykarz.nl/get-sharepoint-tenant-app-catalog-url-rest-api/?

VelinGeorgiev commented:
@waldekmastykarz , I just pasted what we had in the tenant get appcatalog command, but I can change it, should not be a problem. Would you like me to change it in the app catalog command as well. I can create a shared method and use it for both.

waldekmastykarz commented:
If you could first verify, that it indeed works when the app catalog has not yet been indexed, then let's do that indeed.

VelinGeorgiev commented:
@waldekmastykarz , here is what I have found. 

Search, Client.svc and OData do not have advantage over each other. Approx. time for all is like 5-10 mins after the app catalog is specified.

- Search requires crawler to index the app catalog first. 
- Client.svc and the OData services try to get data from the tenant settings and work in a similar way, but there seems to be a cache expiry interval per web api instance.

The search approach in some cases can index the app catalog faster, but in is not guaranteed. This is because there seem to be a cache on the front ends servicing the OData and Client.svc services that has to be invalidated per web api instance. When I hit the OData enpoint I randomly get NULL or valid url meaning that I am hitting different api instances some of which have not invalidated the cache. As result if the rest apis are scaled to 3 api instances the cache can be invalidated to 2 of 3, but if  you hit the 3rd you get NULL result.

Anyway, we are fine in 5-10 minutes after the app catalog is specified.

**Please let me know if I should change it to use the REST endpoint?**

waldekmastykarz commented:
Thanks for the detailed findings @VelinGeorgiev. Let's switch to REST for consistency and simplicity. Sorry for the last minute additional work this is causing 😊

VelinGeorgiev commented:
No worries @waldekmastykarz. I wanted to simplified anyway.

waldekmastykarz commented:
In the `deploy`, `get` and `remove` commands, why not use the existing `appCatalogUrl` option for specifying the URL of the site collection app catalog? Right now we have two options for the same purpose but one exclusive to tenant app catalog and the other to the site collection app catalog. Using the `appCatalogUrl` option for all cases and defining the scope using the `scope` option would be easier, don't you think?

If we'll go with that approach, then I'd suggest consistently using the `appCatalogUrl` option in all app commands.

VelinGeorgiev commented:
@waldekmastykarz, I am confused now. Have we not discussed that up in the same conversation? See: https://github.com/pnp/office365-cli/pull/653#issuecomment-439182529

Could you please give me two examples (code snippets) of how the `get` command would look like when scope sitecollection and when scope is tenant?

waldekmastykarz commented:
Sorry, my bad. Seems like I changed my mind without considering what we discussed. Let's keep everything as-is. Sorry for the confusion.

waldekmastykarz commented on src/o365/spo/commands/app/app-remove.ts:
```diff
@@ -173,12 +168,20 @@ class AppRemoveCommand extends SpoCommand {
 
       if (args.options.appCatalogUrl) {
-        const isValidSharePointUrl: string | boolean = SpoCommand.isValidSharePointUrl(args.options.appCatalogUrl);
+        const isValidSharePointUrl: string | boolean = SpoAppBaseCommand.isValidSharePointUrl(args.options.appCatalogUrl);
```
Since `isValidSharePointUrl` is defined in the `SpoCommand` class, it would be clearer to use that class, instead of `SpoAppBaseCommand`.

waldekmastykarz commented on src/o365/spo/commands/app/app-remove.ts:
```diff
@@ -173,12 +168,20 @@ class AppRemoveCommand extends SpoCommand {
+      }
+
+      if (args.options.siteUrl) {
+        return SpoAppBaseCommand.isValidSharePointUrl(args.options.siteUrl);
```
Since `isValidSharePointUrl` is defined in the `SpoCommand` class, it would be clearer to use that class, instead of `SpoAppBaseCommand`.

waldekmastykarz commented on src/o365/spo/commands/app/app-retract.ts:
```diff
@@ -193,7 +152,7 @@ class AppDeployCommand extends SpoCommand {
 
       if (args.options.appCatalogUrl) {
-        return SpoCommand.isValidSharePointUrl(args.options.appCatalogUrl);
+        return SpoAppBaseCommand.isValidSharePointUrl(args.options.appCatalogUrl);
```
Since `isValidSharePointUrl` is defined in the `SpoCommand` class, it would be clearer to use that class instead of `SpoAppBaseCommand`

waldekmastykarz commented:
@VelinGeorgiev you are right and that's what I've been thinking as well. What confused me initially was the piece related to removing the list name from the URL but now I get that it's meant more as a precaution in case someone included it. Either way, it would make perfect sense to have one option to specify the app catalog URL no matter if it's tenant or site collection and then use the scope option to specify what kind of app catalog the URL is pointing to.

As I mentioned, I've done some fixes on your PR. It's not merged in the dev branch yet, as I noticed some tests to be broken and which I was just fixing. Since I'm working on the PR, I could either take what you suggested above and change in the commands or I could push the temp branch I'm working on right now and you could apply the fix there (with or without fixing the broken tests). I caused enough trouble so far going back and forth, so I'd be totally fine with fixing the mess myself 😊

VelinGeorgiev commented:
@waldekmastykarz, additional suggestion:
Another area of change can be to completely drop the tenant app catalog url as option because you can have only one active at a time and we can auto-discover it. Then the would not be a need to have --appCatalogUrl, but just --siteUrl option when the scope is site. 
Here is my very aggressive suggestion: 

Tenant:
```
o365 spo app list 
```
Sitecollection: 

```
o365 spo app list --siteCatalog https://contoso.sharepoint.com/sites/velindev
```
or
```
o365 spo app list --siteCatalog https://contoso.sharepoint.com/sites/velindev/AppCatalog
```

waldekmastykarz commented:
I'd prefer to keep the appCatalogUrl for two reasons:

- dropping it means a breaking change, means a major release, which I'd prefer to avoid for such a relatively small change
- when you build a script, discovering the URL of the app catalog once and passing it, speeds up the script, also it's a necessity in case you have a new tenant and can't get the URL from the API

VelinGeorgiev commented:
Good points. Let's then NOT go so extreme. :)
